### PR TITLE
distsql: added local and final stage aggregation for STDDEV/VARIANCE

### LINKED ISSUE #14351: distsql: support local aggregation in STDDEV, VARIANCE
Most aggregation functions support a local stage that performs an initial aggregation for the values on each node. For example, for SUM we can take the sum of all values on each node and sum those to get the final result:
image

For AVG, STDDEV, VARIANCE we don't do this; we need to extend the infrastructure to allow more than one intermediary value per function, with a final rendering expression. For AVG, we would have two values (sum and count), and after the final aggregation we would divide them to get the average. Similar solutions exist for STDDEV and VARIANCE.

This would be a great started project for someone because it touches a relatively small surface of distsql.

Update: we now support AVG and there is infrastructure for multiple intermediary values per function.

### PR DESCRIPTION
**Pull request by richardwu**
Fixes #14351

Local/final aggregation for STDDEV/VARIANCE is slightly non-trivial because of how standard 
deviation and variance is calculated. 

The [naive two-pass approach using the sum of squared differences formula](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Two-pass_algorithm) requires the global average to be known, causing local/final aggregation to be irrelevant.

The [second naive approach using the sum of squared values and sum of values](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Na.C3.AFve_algorithm) that can employ local/final aggregation is numerically unstable and can lead to catastrophic cancellation.

Ultimately, the [Knuth/Welford algorithm for variance](https://en.wikipedia.org/wiki/Algorithms_for_calculating_variance#Online_algorithm) allows us to perform local/final aggregation for STDDEV/VARIANCE under distsql. This works for local aggregation (which we currently use for non-distsql calculation of STDDEV/VARIANCE). For final aggregation, we must take the sum of Knuth/Welford squared differences. Refer to https://www.johndcook.com/blog/skewness_kurtosis and our adaptation for `NumericStats` (https://github.com/cockroachdb/cockroach/pull/17728).

### CODE REVIEW

File: pkg/sql/distsqlplan/aggregator_funcs.go [hunk 1]
```diff
@@ -224,6 +224,41 @@ var DistAggregationTable = map[distsqlrun.AggregatorSpec_Func]DistAggregationInf
...
+	//
+	// At a high level, this is analogous to rewriting VARIANCE(x) as SQRDIFF(x)/COUNT(x)
+	// (and STDDEV(x) as sqrt(VARIANCE(x))).
+	distsqlrun.AggregatorSpec_VARIANCE: {
+		LocalStage: []distsqlrun.AggregatorSpec_Func{
+			distsqlrun.AggregatorSpec_SQRDIFF,
+			distsqlrun.AggregatorSpec_SUM,
+			distsqlrun.AggregatorSpec_COUNT,
+		},
+		FinalStage: []FinalStageInfo{
```
Comment (richardwu) [inline pkg/sql/distsqlplan/aggregator_funcs.go]: Point of interest: Instead of have a `SUM_SQDIFFS` and `SUM_INT` (for `COUNT`) stage for `VARIANCE` and `STDDEV` then tailoring a `FinalRendering` stage specific to each, I had to use a specific `FINAL_(VARIANCE|STDDEV)` aggregation stage. I ran into trailing zero issues with `STDDEV` and taking the `Pow(0.5)` of a `SUM_SQDIFFS` decimal value.

File: pkg/sql/parser/aggregate_builtins.go [hunk 2]
```diff
@@ -813,6 +857,282 @@ func (a *intervalSumAggregate) Result() (Datum, error) {
...
+	// is processed, some number of trailing zeros could be added to the
+	// output. Remove them so that the results are the same regardless of order.
+	dd.Decimal.Reduce(&dd.Decimal)
+	return dd, nil
+}
+
+// Close is part of the AggregateFunc interface.
+func (a *decimalSumSqrDiffsAggregate) Close(context.Context) {}
+
+// TODO(richardwu): refactor (int|decimal)VarianceAggregate to varianceAggregate that
```
Comment (richardwu) [inline pkg/sql/parser/aggregate_builtins.go]: There is a lot of copied code here since we needed a way to aggregate only the sum of squared differences in the local stages (as opposed to locally calculated standard deviation/variances).

In a later PR I will consolidate the two implementations of variances (final and non-final) into one that uses the common `sqrDiffAggregate` functions.

**Comment by RaduBerinde** [comment]
This is fantastic work, can't believe how fast you figured everything out!

I don't see the `processors.proto` change that adds `SQRDIFF`, was it left out accidentally? BTW we should also bump DistSQLVersion in `server/distsql.go` with that change (though sync up with @andreimatei on that regarding https://github.com/cockroachdb/cockroach/pull/17767#issuecomment-329941755)

---

Review status: 0 of 4 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7YxuVM-dsinTzlKAR:-Ku7YxuVM-dsinTzlKAS:b-vzsrb4) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/distsqlplan/aggregator_funcs.go#L228)):*
> ```Go
> 	},
> 
> 	// For VARIANCE/STDDEV we need three and two intermediate values in the local and final stage,
> ```

the local stage consists of three aggregations, and the final stage aggregation uses all three values

---

*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 117 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7ZpAyknaep5zETaD6:-Ku7ZpAyknaep5zETaD7:b5py2cz) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/logictest/testdata/logic_test/distsql_agg#L117)):*
> ```
> 22
> 
> # VARIANCE/STDDEV have three local (sqrdiff, sum, and count) and two final stage aggregations.
> ```

one final stage no?

---

*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 132 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7ZbM4Qf35wGFQXMzN:-Ku7ZbM4Qf35wGFQXMzO:bafy7bu) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/logictest/testdata/logic_test/distsql_agg#L132)):*
> ```
> SELECT "URL" FROM [EXPLAIN (DISTSQL) SELECT SUM(a), round(VARIANCE(b), 1) FROM data]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html?eJzElEGL2zAQhe_9FcucUtAhkr3brU4y2wYMrdN6s70UU1RrMIasZSQZWoL_e7F8SGMSuTQQ3yyNn76nN2IO0GiFmXxFC_w7UCDAgEAEBGIgcA8FgdboEq3VZvhlFKTqF_A1gbppOzdsFwRKbRD4AVzt9ggcdvLnHnOUCg0QUOhkvfeQ1tSv0vwWSjoJBLad43eCEsGg6Anozh3PtE5WCJz25N-5SVUZrKTTE-zzy-eVoG-BwPPX_EO62awE86th3389bV-ynf--ZIRdNHLkd402Cg2qE3zR_5fVTZoln358S_I0yZ4-rgQjIiIiHko5NgqNj-7O6K5RQ5VyztNsd_kG0ckN6EItnOHeroVsoQBmuLcLIFoogBnu7QKIFwpghrvMFDtjJEfb6sbiZJqdP3k9TDlUFY4j0erOlPjF6NJjxuXW6_yGQuvGKh0XaTOWBoN_i2lQzE7EdCpmYfIMOgqq47A4vsb3fVD8ECY_XEN-FxQ_hsmP15Dfh3u1nnkm4Uc2ZRf9mz8BAAD__9at3hg=
> ```

Plans look great!

---

*[pkg/sql/parser/aggregate_builtins.go, line 203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7_L-fe0RKJdbgOaIU:-Ku7_L-fe0RKJdbgOaIV:bw0sle5) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L203)):*
> ```Go
> 		makeAggBuiltin([]Type{TypeDecimal, TypeDecimal, TypeInt}, TypeDecimal, newDecimalFinalVarianceAggregate,
> 			"Calculates the variance from the selected locally-computed squared difference values."),
> 		makeAggBuiltin([]Type{TypeDecimal, TypeFloat, TypeInt}, TypeFloat, newFloatFinalVarianceAggregate,
> ```

In what cases would we have a decimal SQDIFF but a float SUM (or vice-versa)? These local aggregations always take the same argument..

---

*[pkg/sql/parser/aggregate_builtins.go, line 1029 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7hPOSG29eh9tdnCN-:-Ku7hPOSG29eh9tdnCN0:b-vhc4p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L1029)):*
> ```Go
> 	delta := mean - a.mean
> 
> 	// Compute the sum of Knuth/Welford sum of squared differences from the mean in a
> ```

[nit] comments are wrapped at 80

---

*[pkg/sql/parser/aggregate_builtins.go, line 1074 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7hW-dnMvefHQJ9aKp:-Ku7hW-dnMvefHQJ9aKq:b-6mede3) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L1074)):*
> ```Go
> }
> 
> func (a *decimalSumSqrDiffsAggregate) Add(_ context.Context, sqrDiffD Datum, otherArgs ...Datum) error {
> ```

[nit] this line look long, make sure you have `https://github.com/cockroachdb/crlfmt` set up or you will encounter lint failures

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-Ku7f_yNuliG_PnyGX4G:b-wq0b7c)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuALY5QxVc6JetsmDgw:-KuALY5QxVc6JetsmDgx:bre8zvz) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/logictest/testdata/logic_test/distsql_agg#L164)):*
> ```
> SELECT "URL" FROM [EXPLAIN (DISTSQL) SELECT SUM(a), AVG(b), SUM(c), AVG(d), STDDEV(a), VARIANCE(b), SUM(a+b+c+d) FROM data]
> ----
> https://cockroachdb.github.io/distsqlplan/decode.html?eJzcVk-Lo04Qvf8-hdTJkIJJq_nnqcJkAsJvMrsmmcsig5suJJCxQ2tglyHffVF3NiqbdiGXkFt1Vz3fq3oF7QekSvIyfucM_G8gAMEBBBcQPEAYQoRw0GrLWaZ0UVIBAvkD_AHCLj0c8-I6QtgqzeB_QL7L9ww-rOPvew45lqwBQXIe7_YlyUHv3mP9k2Scx4AQcipZ-xYJtMhBi1y0yEPLtkn0yen1ye31yYPohKCO-Zkwy-OEwRcn_HdRsyTRnMS5amlabZ5tEj3AKnKK6PFls1z_jstb90_k1fJlvPoazoPFov4NUasRjRqng2fYu9isc7HZc4_HVGnJmmWjxeh05ThWm-e3oBDaGsOn5FrFqDgtguXs_7fVej5_erVpjDRBmp4Tr7MwmC0fn2wSAyQhkMR5AqLkaG7Gw-di0PCBRmjRGC2aoEXTi9NyG9MSt7ivHaLua1-dW3SgQ9R9OeDeogMdou7LAe8WHegQdV8OdPwihJwdVJpx6_X8-5cHxavKMuHqCc7UUW_5i1bbkqY6vpS48kJylldZUR2CtEoVAutgYQQ7DbBogx0zcwe1a0R7ZrB3je6hETwyM4-uYR4bwRMz8-Qa5qnZq0HHmpiXrM0dnf77FQAA__-t0YJ8
> ```

It would be cool if we de-duplicated redundant aggregations like `SUM(@1)` here. Can you file an issue?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520)*

**Comment by richardwu** [comment]
Just to confirm, @andrei should add a comment regarding the additional proto enum values to https://github.com/cockroachdb/cockroach/pull/18522 ?

---

Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 228 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7YxuVM-dsinTzlKAR:-KuHfVjegywNlS7w73Ct:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/distsqlplan/aggregator_funcs.go#L228)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

the local stage consists of three aggregations, and the final stage aggregation uses all three values
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 117 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7ZpAyknaep5zETaD6:-KuHfmeSFdNYVNMY8sLT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/logictest/testdata/logic_test/distsql_agg#L117)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

one final stage no?
</blockquote></details>

Done.

---

*[pkg/sql/logictest/testdata/logic_test/distsql_agg, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuALY5QxVc6JetsmDgw:-KuHft5fcOqHpdZcGwp3:b-jn6wq7) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/logictest/testdata/logic_test/distsql_agg#L164)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

It would be cool if we de-duplicated redundant aggregations like `SUM(@1)` here. Can you file an issue?
</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/18534.

---

*[pkg/sql/parser/aggregate_builtins.go, line 203 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7_L-fe0RKJdbgOaIU:-KuHfyfX6E1iCpzoXG5Y:b7yherm) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L203)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

In what cases would we have a decimal SQDIFF but a float SUM (or vice-versa)? These local aggregations always take the same argument..
</blockquote></details>

Good catch, I'll remove the extraneous ones for now.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1029 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7hPOSG29eh9tdnCN-:-KuHgGiGQ4NgNxviW6k-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L1029)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] comments are wrapped at 80
</blockquote></details>

Done.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1074 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7hW-dnMvefHQJ9aKp:-KuHgJkkCBsSVNVNSHt2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/151dad63dcdf84dd6d103ba420293123374d4a22/pkg/sql/parser/aggregate_builtins.go#L1074)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] this line look long, make sure you have `https://github.com/cockroachdb/crlfmt` set up or you will encounter lint failures
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-KuHfFXVupe8I-8gH_Dl:bxm3d11)*

**Comment by andreimatei** [comment]
I believe Radu's saying that you should bump the version here if you add some
https://github.com/cockroachdb/cockroach/blob/620ab6a3835c46b0b66405fc010dfa54774645ab/pkg/sql/distsqlrun/server.go#L74

Talk to me if the comments there don't make sense.

---

Review status: 0 of 7 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-KuKnvbtz1cUV3jmXVT5:b-tg4gci)*

**Comment by richardwu** [comment]
Could use another :eyes: 

---

Review status: 0 of 9 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-KuLq2-SH0BdRXDGwP47:bagn1k8)*

**Comment by arjunravinarayan** [comment]
Taking over this review. Everything looks good, except for the redundant code in `aggregate_builtins`. I'd refactor that now (as you've suggested doing in a later PR), and then this is good to merge!

---

Reviewed 1 of 2 files at r2, 3 of 6 files at r3, 2 of 3 files at r4, 2 of 2 files at r5.
Review status: 8 of 9 files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7D1mo5Dh8uEYDikSq-r1-241:-KuPamIkEK0t_BQ8UKLB:b-c13ma) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9a097949b7a31d4cf0754a817308dc9a6ad224/pkg/sql/distsqlplan/aggregator_funcs.go#L241)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

Point of interest: Instead of have a `SUM_SQDIFFS` and `SUM_INT` (for `COUNT`) stage for `VARIANCE` and `STDDEV` then tailoring a `FinalRendering` stage specific to each, I had to use a specific `FINAL_(VARIANCE|STDDEV)` aggregation stage. I ran into trailing zero issues with `STDDEV` and taking the `Pow(0.5)` of a `SUM_SQDIFFS` decimal value.
</blockquote></details>

Worth adding a code comment with this information, PR review comments get lost in the sea.

---

*[pkg/sql/distsqlrun/version_history.txt, line 11 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPYcWYFefj6s9FdC20:-KuPYcWYFefj6s9FdC21:b-50u9kr) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/distsqlrun/version_history.txt#L11)):*
> ```
>   - Three new aggregations (SQRDIFF, FINAL_VARIANCE, FINAL_STDDEV) were
>     introduced to support local and final aggregation of the builtin
>     functions STDDEV and VARIANCE. The enum Func under message AggregatorSpec
> ```

Last sentence is unnecessary detail. Do add that older versions won't recognize these aggregations and thus fail.

---

*[pkg/sql/logictest/testdata/logic_test/aggregate, line 858 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPZAA2OYt0fVneP8IO:-KuPZAA2OYt0fVneP8IP:b-vx583l) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/logictest/testdata/logic_test/aggregate#L858)):*
> ```
> NULL
> 
> # Numerical stability test for VARIANCE/STDDEV.
> ```

Nice work! This is good testing.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1133 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7D1mo5Dh8uEYDikSt-r1-1133:-KuPabHwkoWA32kJZJjO:b-2xfr8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9a097949b7a31d4cf0754a817308dc9a6ad224/pkg/sql/parser/aggregate_builtins.go#L1133)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

There is a lot of copied code here since we needed a way to aggregate only the sum of squared differences in the local stages (as opposed to locally calculated standard deviation/variances).

In a later PR I will consolidate the two implementations of variances (final and non-final) into one that uses the common `sqrDiffAggregate` functions.
</blockquote></details>

I'm not super happy about the copied code. Why do we have to wait for a later PR?

---

*[pkg/sql/parser/aggregate_builtins.go, line 999 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPaVXJceUzwPFjKxeE:-KuPaVXJceUzwPFjKxeF:b-mm8342) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/parser/aggregate_builtins.go#L999)):*
> ```Go
> 	}
> 
> 	// Square differenec and sum arguments can either be decimal or float (but not both decimal).
> ```

typo 'differenec'

---

*[pkg/sql/parser/aggregate_builtins.go, line 1085 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPaO1M9gVa0WHV8IJ8:-KuPaO1M9gVa0WHV8IJ9:b-g3xt1c) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/parser/aggregate_builtins.go#L1085)):*
> ```Go
> 	a.ed.Sub(&a.delta, &a.tmpMean, &a.mean)
> 
> 	// Compute the sum of Knuth/Welford sum of squared differences from the mean in a
> ```

Wrap comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-KuPakNqHMRRpJaBwk_I:bp9ybp5)*

**Comment by richardwu** [comment]
Review status: 6 of 9 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/sql/distsqlplan/aggregator_funcs.go, line 241 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7D1mo5Dh8uEYDikSq-r1-241:-KuQrYbxp1T0s1xqTE7Q:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9a097949b7a31d4cf0754a817308dc9a6ad224/pkg/sql/distsqlplan/aggregator_funcs.go#L241)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Worth adding a code comment with this information, PR review comments get lost in the sea.
</blockquote></details>

Done.

---

*[pkg/sql/distsqlrun/version_history.txt, line 11 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPYcWYFefj6s9FdC20:-KuQqsNFRFqCfkyLtY34:b-yxxshw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/distsqlrun/version_history.txt#L11)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Last sentence is unnecessary detail. Do add that older versions won't recognize these aggregations and thus fail.
</blockquote></details>

So older versions (v6s) will never receive (nor generate) v7 protos (re: Version: 7)

v7 servers can still accept v6 protos (re: MinAcceptedVersion: 6)

This is be described here https://github.com/andreimatei/cockroach/blob/b71e34a28ab8a0ea8678101ecd6134c2c50ddd46/pkg/sql/distsqlrun/server.go#L48L77.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1133 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-Ku7D1mo5Dh8uEYDikSt-r1-1133:-KuRcd7rn23p4_GdtWHZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/3d9a097949b7a31d4cf0754a817308dc9a6ad224/pkg/sql/parser/aggregate_builtins.go#L1133)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

I'm not super happy about the copied code. Why do we have to wait for a later PR?
</blockquote></details>

Done.

---

*[pkg/sql/parser/aggregate_builtins.go, line 999 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPaVXJceUzwPFjKxeE:-KuQrVYeGyTygDtye3-N:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/parser/aggregate_builtins.go#L999)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

typo 'differenec'
</blockquote></details>

Done.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1085 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPaO1M9gVa0WHV8IJ8:-KuQrWl7gvEPQ8gZGJU0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/parser/aggregate_builtins.go#L1085)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Wrap comment.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520)*

**Comment by arjunravinarayan** [comment]
Looks good to me, but cc @mjibson to take a look at the changes in `aggregate_builtins` to see if it looks good to him as well.

---

Reviewed 3 of 3 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/distsqlrun/version_history.txt, line 11 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPYcWYFefj6s9FdC20:-KuW9WcsZkFS1-HzGNWp:b-ayd6wc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/distsqlrun/version_history.txt#L11)):*
<details><summary><i>Previously, richardwu (Richard Wu) wrote…</i></summary><blockquote>

So older versions (v6s) will never receive (nor generate) v7 protos (re: Version: 7)

v7 servers can still accept v6 protos (re: MinAcceptedVersion: 6)

This is be described here https://github.com/andreimatei/cockroach/blob/b71e34a28ab8a0ea8678101ecd6134c2c50ddd46/pkg/sql/distsqlrun/server.go#L48L77.
</blockquote></details>

Let me rephrase: I would like your change to say something like this:
"Three new aggregations (...) were introduced to support local and final aggregation of the builtin functions STDDEV and VARIANCE. These new aggregations would be unrecognized by a server running old versions, hence the version bump. However, a server running v7 can still process all plans from servers running version 6,  thus the MinAcceptedVersion is kept at 6."

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-KuWE8hqvmayxnJPepdE:bc6ma46)*

**Comment by mjibson** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1063 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuXYTYWWaDTQIgYIOLz:-KuXYTYXeC4AaLCZToRU:b-ijhtt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/7954107c3ae81b57673b93b84aad1d12c9bcf1b9/pkg/sql/parser/aggregate_builtins.go#L1063)):*
> ```Go
> 	// https://github.com/cockroachdb/cockroach/pull/17728.
> 	a.sqrDiff += sqrDiff + delta*delta*float64(count)*float64(a.count)/float64(a.count+count)
> 	a.count += count
> ```

Both here and on the line above, use addWithOverflow (https://github.com/cockroachdb/cockroach/blob/62fae528ceec1f4c3d3470f2a1e2bb9fad3b5fd8/pkg/sql/parser/eval.go#L323) to ensure the ints can add safely.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520)*

**Comment by richardwu** [comment]
Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/sql/distsqlrun/version_history.txt, line 11 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuPYcWYFefj6s9FdC20:-Ku__-iTrSZlg7ZQGQN-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ece397613450beae16171850e03dfed2c7cd87e/pkg/sql/distsqlrun/version_history.txt#L11)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

Let me rephrase: I would like your change to say something like this:
"Three new aggregations (...) were introduced to support local and final aggregation of the builtin functions STDDEV and VARIANCE. These new aggregations would be unrecognized by a server running old versions, hence the version bump. However, a server running v7 can still process all plans from servers running version 6,  thus the MinAcceptedVersion is kept at 6."

</blockquote></details>

Done.

---

*[pkg/sql/parser/aggregate_builtins.go, line 1063 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-KuXYTYWWaDTQIgYIOLz:-Ku__11k9C9iS3QFiWFr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7954107c3ae81b57673b93b84aad1d12c9bcf1b9/pkg/sql/parser/aggregate_builtins.go#L1063)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Both here and on the line above, use addWithOverflow (https://github.com/cockroachdb/cockroach/blob/62fae528ceec1f4c3d3470f2a1e2bb9fad3b5fd8/pkg/sql/parser/eval.go#L323) to ensure the ints can add safely.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520)*

**Comment by mjibson** [comment]
:lgtm:

---

Review status: 7 of 9 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18520#-:-Ku_imiEDSPVMtdMXXGR:bnfp4nl)*
